### PR TITLE
ARROW-14822: [C++] Implement floor/ceil/round for temporal objects

### PR DESCRIPTION
**Pull request by rok**
This is to resolve [ARROW-14822](https://issues.apache.org/jira/browse/ARROW-14822).

### CODE REVIEW

File: cpp/src/arrow/compute/api_scalar.h [hunk 1]
```diff
@@ -108,6 +108,32 @@ class ARROW_EXPORT RoundToMultipleOptions : public FunctionOptions {
   RoundMode round_mode;
 };
 
+class ARROW_EXPORT RoundTemporalOptions : public FunctionOptions {
+ public:
+  enum Ambiguous { AMBIGUOUS_RAISE, AMBIGUOUS_EARLIEST, AMBIGUOUS_LATEST };
+  enum Nonexistent { NONEXISTENT_RAISE, NONEXISTENT_EARLIEST, NONEXISTENT_LATEST };
+
+  explicit RoundTemporalOptions(std::string frequency = "day",
```
Comment (lidavidm) [inline cpp/src/arrow/compute/api_scalar.h:116]: I realize this is just a draft :) that said, should be an enum as well?

File: cpp/src/arrow/compute/api_scalar.cc [hunk 2]
```diff
@@ -228,6 +268,12 @@ static auto kRoundOptionsType = GetFunctionOptionsType<RoundOptions>(
 static auto kRoundToMultipleOptionsType = GetFunctionOptionsType<RoundToMultipleOptions>(
     DataMember("multiple", &RoundToMultipleOptions::multiple),
     DataMember("round_mode", &RoundToMultipleOptions::round_mode));
+static auto kRoundTemporalOptionsType = GetFunctionOptionsType<RoundTemporalOptions>(
```
Comment (lidavidm) [inline cpp/src/arrow/compute/api_scalar.cc:271]: nit: another contributor just alphabetized these, so maybe try to preserve that (same goes below).

File: cpp/src/arrow/compute/api_scalar.h (hunk 1 above)
Comment (rok) [reply cpp/src/arrow/compute/api_scalar.h:116]: Too soon David :)

File: cpp/src/arrow/compute/api_scalar.cc (hunk 2 above)
Comment (rok) [reply cpp/src/arrow/compute/api_scalar.cc:271]: Oh yeah, will do!

File: cpp/src/arrow/compute/api_scalar.h (hunk 1 above)
Comment (rok) [reply cpp/src/arrow/compute/api_scalar.h:116]: Thanks for the feedback though! will do.

File: cpp/src/arrow/compute/api_scalar.cc (hunk 2 above)
Comment (rok) [reply cpp/src/arrow/compute/api_scalar.cc:337]: Done.

**Comment by rok** [comment]
This is getting review ready, couple of issues remaining:
* `RoundTemporal`, `CeilTemporal` and `FloorTemporal` could maybe be templated from a single struct
* `Nonexistent` in `RoundTemporalOptions` might be unnecessary as we assume timestamps are in UTC so they always exist. Probably holds for `Ambiguous` too.
* Test data needs some work, especially pre-1970 and edge-cases.

**Comment by rok** [comment]
@lidavidm @jorisvandenbossche could you take a quick glance to see if something needs to be done fundamentally different here? If not I hope to have this review ready tomorrow.

**Comment by lidavidm** [comment]
> * `Nonexistent` in `RoundTemporalOptions` might be unnecessary as we assume timestamps are in UTC so they always exist. Probably holds for `Ambiguous` too.

Would we want to eventually support rounding in a timezone? (That can be deferred since I can see that getting tricky.) e.g. rounding to the hour may differ in a timezone that is xx:30 offset from UTC.

**Comment by rok** [comment]
> > * `Nonexistent` in `RoundTemporalOptions` might be unnecessary as we assume timestamps are in UTC so they always exist. Probably holds for `Ambiguous` too.
> 
> Would we want to eventually support rounding in a timezone? (That can be deferred since I can see that getting tricky.) e.g. rounding to the hour may differ in a timezone that is xx:30 offset from UTC.

I've been thinking about that exact situation (xx:30) and I'm not sure yet how to tackle it but I'd like to in this ticket.

**Comment by lidavidm** [comment]
I only took a quick scan but I don't see anything fundamentally wrong. 

- It gets confusing with all the `chrono` types but we should stick to UpperCamelCase for Arrow functions/methods
- Do we want QUARTER in addition to SEASON?

**Comment by rok** [comment]
> * Do we want QUARTER in addition to SEASON?

I had it originally but it felt redundant as it's just `RoundTemporal(arg, RoundTemporal(unit=month, multiple=3))`. I'll add it back as it will probably get requested.

**Comment by lidavidm** [comment]
Ah, that is true - we could just document it.

**Comment by rok** [comment]
@lidavidm - I think this is ready for first round of review.
I've decided to remove `Nonexistent` and `Ambiguous`.
I'm working out some issues with rounding multiple seasons but other than that I think it's good to review. I'll also try adding `origin` to set the "zero time" in one of the future commits.

**Comment by rok** [comment]
@pitrou any thoughts on this?

File: cpp/src/arrow/compute/api_scalar.h [hunk 3]
```diff
@@ -90,6 +90,36 @@ class ARROW_EXPORT RoundOptions : public FunctionOptions {
...
+  NANOSECOND,
+  MICROSECOND,
+  MILLISECOND,
+  SECOND,
+  MINUTE,
+  HOUR,
+  DAY,
+  WEEK,
+  MONTH,
+  SEASON,
```
Comment (lidavidm) [inline cpp/src/arrow/compute/api_scalar.h:103]: Can we define exactly what SEASON is?

**Review by lidavidm — COMMENTED** [review]
Overall looks good to me, taking into account the TODOs.

File: cpp/src/arrow/compute/api_scalar.h (hunk 3 above)
Comment (rok) [reply cpp/src/arrow/compute/api_scalar.h:103]: I'm not exactly sure myself. Here it [looks like quarter](https://github.com/tidyverse/lubridate/blob/9cece5680d0280c5cd1ccdbc9309ccb4724b3a34/R/round.r#L199-L201) but from looking at output it seems to be shifted a month back (Winter runs Dec 1 - March 1 etc) and it's not stable - first season start moves depending on the multiple.
I'll just add quarter and we can put this in if desired later.

File: python/pyarrow/_compute.pyx [hunk 4]
```diff
@@ -774,6 +774,48 @@ class RoundOptions(_RoundOptions):
...
+        return CCalendarUnit_DAY
+    elif unit == "week":
+        return CCalendarUnit_WEEK
+    elif unit == "month":
+        return CCalendarUnit_MONTH
+    elif unit == "season":
+        return CCalendarUnit_SEASON
+    elif unit == "year":
+        return CCalendarUnit_YEAR
+    _raise_invalid_function_option(unit, "Calendar unit")
```
[the remark this answers is not in the document]
Comment (rok) [reply python/pyarrow/_compute.pyx:800]: There is pandas shorthand and R shorthand. I'll check if it makes sense to parse them in c++.

File: cpp/src/arrow/compute/api_scalar.h [hunk 5]
```diff
@@ -90,6 +90,36 @@ class ARROW_EXPORT RoundOptions : public FunctionOptions {
...
+  explicit RoundTemporalOptions(int multiple = 1, CalendarUnit unit = CalendarUnit::DAY,
+                                int origin = 0);
+  constexpr static char const kTypeName[] = "RoundTemporalOptions";
+  static RoundTemporalOptions Defaults() { return RoundTemporalOptions(); }
+
+  /// Number of units to round to
+  int multiple;
+  /// The unit used for rounding of time
+  CalendarUnit unit;
+  /// Origin time of rounding
```
[the remark this answers is not in the document]
Comment (rok) [reply cpp/src/arrow/compute/api_scalar.h:119]: Yeah origin would be important for "political units" (minute and larger). I'll document it better.

File: cpp/src/arrow/compute/kernels/scalar_temporal_unary.cc [hunk 6]
```diff
@@ -452,6 +470,341 @@ struct Nanosecond {
   }
 };
 
+// ----------------------------------------------------------------------
+// Round temporal values to given frequency
+
+Result<ValueDescr> ResolveRoundTemporalOutput(KernelContext* ctx,
```
[the remark this answers is not in the document]
Comment (rok) [reply cpp/src/arrow/compute/kernels/scalar_temporal_unary.cc:479]: I was thinking to maybe change unit when rounding. Going to FirstType for now.

File: cpp/src/arrow/compute/kernels/scalar_temporal_unary.cc [hunk 7]
```diff
@@ -128,6 +137,18 @@ struct TemporalComponentExtractWeek
   }
 };
 
+template <template <typename...> class Op, typename Duration, typename InType,
+          typename OutType>
+struct TemporalComponentExtractWithOptions
```
[the remark this answers is not in the document]
Comment (rok) [reply cpp/src/arrow/compute/kernels/scalar_temporal_unary.cc]: Done.

File: python/pyarrow/tests/test_compute.py [hunk 8]
```diff
@@ -1848,6 +1851,60 @@ def test_assume_timezone():
...
+    "millisecond": "L",
+    "second": "s",
+    "minute": "min",
+    "hour": "H",
+    "day": "D"
+}
+
+
+@pytest.mark.parametrize('unit', units)
+def test_round_temporal(unit):
```
[the remark this answers is not in the document]
Comment (rok) [reply python/pyarrow/tests/test_compute.py:1874]: Done.

**Comment by jorisvandenbossche** [comment]
> I've decided to remove `Nonexistent` and `Ambiguous`.

(didn't yet look at the code) Did you then also remove support rounding timestamps with a timezone? Or how does rounding tz-aware timestamps work?

It's also not really clear to me what `origin` would do. Can you explain with an example what it would do? (I also don't see it in lubridate or pandas?)

**Comment by rok** [comment]
> > I've decided to remove `Nonexistent` and `Ambiguous`.
> 
> (didn't yet look at the code) Did you then also remove support rounding timestamps with a timezone? Or how does rounding tz-aware timestamps work?

At the moment some of the code (ns, us, ms, s, m, h, d, w) is UTC based even when timezones are present. UTC always converts to a local time so nonexistent is not an issue while ambiguous could be.
Month, quarter, year are currently rounded in local time and integer multiples of those "units" are less problematic because beginnings of months seem to always exist. Perhaps this needs more study.

> It's also not really clear to me what `origin` would do. Can you explain with an example what it would do? (I also don't see it in lubridate or pandas?)

Origin is meant like the origin in Euclidean space. Let's say you want to floor today's date to `100year`:
```python
floor("2021-12-22", unit="year", multiple=100, origin="1970-01-01") == "1970-01-01"
floor("2021-12-22", unit="year", multiple=100, origin="0-01-01") == "2000-01-01"
```

R's [clock](https://clock.r-lib.org/reference/date-and-date-time-rounding.html) has this concept.
It's also a "replacement" for `week_starts_monday`.
e.g.: `1970-01-01` was a Thursday so `floor(t, unit="week", multiple=1, origin="1970-01-05")` is equivalent to `floor(t, unit="week", multiple=1, week_starts_monday=True)`

**Comment by rok** [comment]
I think tests are now mostly in order. Remaining work:
* origin
* timezone handling (nonexistent, ambiguous)
* shorthand

Am I forgetting something?

Which of these could be pushed out of this scope? I'm thinking shorthand and perhaps perhaps timezone handling.
@lidavidm @jorisvandenbossche

**Comment by jorisvandenbossche** [comment]
> At the moment some of the code (ns, us, ms, s, m, h, d, w) is UTC based even when timezones are present. UTC always converts to a local time so nonexistent is not an issue while ambiguous could be.

In pandas we always round in local time (some discussion about this in https://github.com/pandas-dev/pandas/issues/37592), and then you can actually run into nonexistent times as well. 
The reason we do this (with its limitations ..), is because not every timezone has an exactly rounded / hourly offset, and so because of that rounding can give wrong results. For example, in Australia in summer there is a UTC+10:30 timezone:

```python
>>> ts = pd.Timestamp("2021-07-01 12:17:00", tz="Australia/Lord_Howe")
>>>ts
Timestamp('2021-07-01 12:17:00+1030', tz='Australia/Lord_Howe')
>>> ts.round("H")
Timestamp('2021-07-01 12:00:00+1030', tz='Australia/Lord_Howe')

# with this PR with pyarrow
>>> arr = pa.array([ts])
>>> pc.round_temporal(arr, unit="hour")
<pyarrow.lib.TimestampArray object at 0x7f743c72dc40>
[
  2021-07-01 02:00:00.000000
]
>>> pc.round_temporal(arr, unit="hour")[0]
<pyarrow.TimestampScalar: datetime.datetime(2021, 7, 1, 12, 30, tzinfo=<DstTzInfo 'Australia/Lord_Howe' +1030+10:30:00 STD>)>
```

So with pyarrow that actually gives the "wrong" value now because of relying on rounding in UTC. 

As I mentioned in the above-linked pandas issue, rounding in local time _does_ have some serious drawbacks (eg those annoying nonexistent times that actually should be possible to round properly without error), so it would be nice if we could somehow detect the cases where it is fine to do this. But doing that robustly might be quite complex ...

**Comment by jorisvandenbossche** [comment]
I suppose a simpler example is actually also rounding to a multiple of hours in a normal timezone. For example for me, the current UTC offset is +1, so rounding to two hours will give a different result when doing it in UTC:

```python
>>> ts = pd.Timestamp("2021-12-23 12:17:00", tz="Europe/Brussels")
>>> ts
Timestamp('2021-12-23 12:17:00+0100', tz='Europe/Brussels')
# with pandas I get 12:00
>>> ts.round("2H")
Timestamp('2021-12-23 12:00:00+0100', tz='Europe/Brussels')

>>> arr = pa.array([ts])
>>> pc.round_temporal(arr, unit="hour", multiple=2)
<pyarrow.lib.TimestampArray object at 0x7f743c74cc40>
[
  2021-12-23 12:00:00.000000
]
# but with arrow I get 12:00 in UTC, so 13:00 in local time
>>> pc.round_temporal(arr, unit="hour", multiple=2)[0]
<pyarrow.TimestampScalar: datetime.datetime(2021, 12, 23, 13, 0, tzinfo=<DstTzInfo 'Europe/Brussels' CET+1:00:00 STD>)>
```

**Comment by rok** [comment]
> > At the moment some of the code (ns, us, ms, s, m, h, d, w) is UTC based even when timezones are present. UTC always converts to a local time so nonexistent is not an issue while ambiguous could be.
> 
> In pandas we always round in local time (some discussion about this in [pandas-dev/pandas#37592](https://github.com/pandas-dev/pandas/issues/37592)), and then you can actually run into nonexistent times as well.

I'm happy to hear pandas rounds in local time! (For some reason I assumed it doesn't).

> The reason we do this (with its limitations ..), is because not every timezone has an exactly rounded / hourly offset, and so because of that rounding can give wrong results. For example, in Australia in summer there is a UTC+10:30 timezone:

Yeah, David raised this point too, I'll try to address it and get back.

**Comment by jorisvandenbossche** [comment]
If you are looking for reducing scope of this PR, we could also defer rounding tz-aware timestamps for later. But then for this PR we should simply disallow it (raise an error) IMO, instead of rounding in UTC.

**Comment by lidavidm** [comment]
Agreed, if TZ handling is complex we can defer it, but we should error like other kernels instead of returning an unexpected result.

I think the shorthand is not necessary either. I was thinking that could be done in the bindings instead of C++, in any case.

**Comment by jorisvandenbossche** [comment]
Yes, I agree that adding frequency shorthands should maybe not be done in C++ (or is at least not important for this PR)

File: python/pyarrow/tests/test_compute.py [hunk 9]
```diff
@@ -1848,6 +1849,64 @@ def test_assume_timezone():
...
+
+# TODO: We should test on windows once ARROW-13168 is resolved.
+@pytest.mark.pandas
+@pytest.mark.skipif(sys.platform == 'win32',
+                    reason="Timezone database is not available on Windows yet")
+@pytest.mark.parametrize('unit', units)
+def test_round_temporal(unit):
+    # timezones = ["UTC", "US/Central", "Asia/Kolkata",
+    #              "Etc/GMT-4", "Etc/GMT+4", "Australia/Broken_Hill"]
+    # timezones = ["UTC", "Etc/GMT+4"]
```
Comment (lidavidm) [inline python/pyarrow/tests/test_compute.py:1880]: Are these lines meant to be commented out still? (And below)

File: cpp/src/arrow/compute/kernels/scalar_temporal_test.cc [hunk 10]
```diff
@@ -1324,6 +1361,555 @@ TEST_F(ScalarTemporalTest, TestTemporalDifferenceZoned) {
...
+  CheckScalarUnary(op, unit, times, unit, round_15_seconds, &round_to_15_seconds);
+  CheckScalarUnary(op, unit, times, unit, round_15_minutes, &round_to_15_minutes);
+  CheckScalarUnary(op, unit, times, unit, round_15_hours, &round_to_15_hours);
+  CheckScalarUnary(op, unit, times, unit, round_15_days, &round_to_15_days);
+  CheckScalarUnary(op, unit, times, unit, round_15_weeks, &round_to_15_weeks);
+  CheckScalarUnary(op, unit, times, unit, round_15_months, &round_to_15_months);
+  CheckScalarUnary(op, unit, times, unit, round_15_quarters, &round_to_15_quarters);
+  CheckScalarUnary(op, unit, times, unit, round_15_years, &round_to_15_years);
+
+  RoundTemporalOptions round_to_2_hours = RoundTemporalOptions(2, CalendarUnit::HOUR);
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:1906]: We need to #ifdef these out on Windows right?

File: python/pyarrow/tests/test_compute.py (hunk 9 above)
Comment (rok) [reply python/pyarrow/tests/test_compute.py]: They are commented out to please the CI. Still working on the C++ logic.

File: cpp/src/arrow/compute/kernels/scalar_temporal_test.cc (hunk 10 above)
Comment (rok) [reply cpp/src/arrow/compute/kernels/scalar_temporal_test.cc]: I think the whole bottom of the file is already and it might not show in the diff.

Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/scalar_temporal_test.cc]: Oops, I didn't expand it out enough. Thanks.

File: cpp/src/arrow/compute/kernels/scalar_temporal_test.cc [hunk 11]
```diff
@@ -342,6 +342,39 @@ class ScalarTemporalTest : public ::testing::Test {
...
+      RoundTemporalOptions(15, CalendarUnit::SECOND);
+  RoundTemporalOptions round_to_15_minutes =
+      RoundTemporalOptions(15, CalendarUnit::MINUTE);
+  RoundTemporalOptions round_to_15_hours = RoundTemporalOptions(15, CalendarUnit::HOUR);
+  RoundTemporalOptions round_to_15_days = RoundTemporalOptions(15, CalendarUnit::DAY);
+  RoundTemporalOptions round_to_15_weeks = RoundTemporalOptions(15, CalendarUnit::WEEK);
+  RoundTemporalOptions round_to_15_months = RoundTemporalOptions(15, CalendarUnit::MONTH);
+  RoundTemporalOptions round_to_15_quarters =
+      RoundTemporalOptions(15, CalendarUnit::QUARTER);
+  RoundTemporalOptions round_to_15_years = RoundTemporalOptions(15, CalendarUnit::YEAR);
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:377]: Should we have a test for the `origin` option too?

Comment (rok) [reply cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:377]: I've take origin out for now sorry for the back and forth.

**Comment by rok** [comment]
@lidavidm @jorisvandenbossche this now rounds time since epoch in local time ([see here](https://github.com/apache/arrow/blob/b93f0758cd24f23f20ee7b0939ff409f213a9df7/cpp/src/arrow/compute/kernels/scalar_temporal_unary.cc#L502-L523)). It works ok outside of DST and discussible within. It's a different approach to what e.g. Pandas does.
I'm going to look into a better approach for this.

File: python/pyarrow/tests/test_compute.py [hunk 12]
```diff
@@ -1899,6 +1900,71 @@ def test_assume_timezone():
...
+    }
+    ta = pa.array(ts)
+
+    for value in values:
+        frequency = str(value) + unit_shorthand[unit]
+        options = pc.RoundTemporalOptions(value, unit)
+
+        result = pc.ceil_temporal(ta, options=options).to_pandas()
+        expected = ts.dt.ceil(frequency)
+        if unit != "nanosecond":
```
Comment (lidavidm) [inline python/pyarrow/tests/test_compute.py:1921]: What's the difference with Pandas here?

File: python/pyarrow/tests/test_compute.py [hunk 13]
```diff
@@ -1899,6 +1900,71 @@ def test_assume_timezone():
...
+                    reason="Timezone database is not available on Windows yet")
+@pytest.mark.parametrize('unit', ("nanosecond", "microsecond", "millisecond",
+                                  "second", "minute", "hour", "day"))
+@pytest.mark.pandas
+def test_round_temporal(unit):
+    values = (1, 2, 3, 4, 5, 6, 7, 10, 15, 24, 60, 250, 500, 750)
+    timestamps = [
+        # "1899-04-18 01:57:09.190202880",
+        # "1899-09-12 07:03:30.080325120",
+        # "1904-06-21 20:55:36.493869056",
```
Comment (lidavidm) [inline python/pyarrow/tests/test_compute.py:1944]: Are these still meant to be commented out?

File: cpp/src/arrow/compute/kernels/scalar_temporal_unary.cc [hunk 14]
```diff
@@ -452,6 +474,296 @@ struct Nanosecond {
...
+      case compute::CalendarUnit::MONTH: {
+        auto t0 = localizer_.template ConvertTimePoint<Duration>(arg);
+        year_month_day ymd =
+            GetFlooredYmd<Duration, Localizer>(arg, options.multiple, localizer_);
+
+        auto f = localizer_.ConvertDays(ymd.year() / ymd.month() / 1);
+        ymd += months{options.multiple};
+        auto c = localizer_.ConvertDays(ymd.year() / ymd.month() / 1);
+
+        t = (t0 - f > c - t0) ? c.time_since_epoch() : f.time_since_epoch();
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/scalar_temporal_unary.cc:732]: Just curious, should this be `>=`?

File: cpp/src/arrow/compute/kernels/scalar_temporal_test.cc [hunk 15]
```diff
@@ -1324,6 +1357,578 @@ TEST_F(ScalarTemporalTest, TestTemporalDifferenceZoned) {
...
+
+  auto unit = timestamp(TimeUnit::NANO, "UTC");
+//  CheckScalarUnary(op, unit, times, unit, ceil_1_nanosecond, &round_to_1_nanoseconds);
+//  CheckScalarUnary(op, unit, times, unit, ceil_1_microsecond, &round_to_1_microseconds);
+//  CheckScalarUnary(op, unit, times, unit, ceil_1_millisecond, &round_to_1_milliseconds);
+//  CheckScalarUnary(op, unit, times, unit, ceil_1_second, &round_to_1_seconds);
+//  CheckScalarUnary(op, unit, times, unit, ceil_1_minute, &round_to_1_minutes);
+//  CheckScalarUnary(op, unit, times, unit, ceil_1_hour, &round_to_1_hours);
+//  CheckScalarUnary(op, unit, times, unit, ceil_1_day, &round_to_1_days);
+//  CheckScalarUnary(op, unit, times, unit, ceil_1_weeks, &round_to_1_weeks);
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:1517]: There's commented out tests here too.

File: python/pyarrow/tests/test_compute.py (hunk 12 above)
Comment (rok) [reply python/pyarrow/tests/test_compute.py]: There was a difference in the `ceil` function. It's ok now.

File: cpp/src/arrow/compute/kernels/scalar_temporal_unary.cc (hunk 14 above)
Comment (rok) [reply cpp/src/arrow/compute/kernels/scalar_temporal_unary.cc]: I'm not sure how to round half by default. Switching to up (`>=`).

File: python/pyarrow/tests/test_compute.py (hunk 13 above)
Comment (rok) [reply python/pyarrow/tests/test_compute.py:1943]: There are some odd changes in timezone offsets that are causing problems. I'll try changing `ceil` to match behaviour.

File: cpp/src/arrow/compute/api_scalar.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (rok) [reply cpp/src/arrow/compute/api_scalar.h]: Dropped origin for now.

File: python/pyarrow/tests/test_compute.py (hunk 13 above)
Comment (rok) [reply python/pyarrow/tests/test_compute.py:1945]: This is the kind of error I'm seeing for these three timestamps:

expected:
```
Timestamp('1899-04-19 00:31:50+0553', tz='Asia/Kolkata')
Timestamp('1899-09-13 00:31:50+0553', tz='Asia/Kolkata')
Timestamp('1904-06-22 23:59:50+0521', tz='Asia/Kolkata')
````

actual:
```
Timestamp('1899-04-19 00:00:00+0553', tz='Asia/Kolkata'),
Timestamp('1899-09-13 00:00:00+0553', tz='Asia/Kolkata'),
Timestamp('1904-06-23 00:00:00+0521', tz='Asia/Kolkata')
```

Comment (rok) [reply python/pyarrow/tests/test_compute.py:1945]: https://en.wikipedia.org/wiki/Calcutta_Time

Comment (lidavidm) [reply python/pyarrow/tests/test_compute.py:1945]: Oh boy. So, if I'm reading this right, it's rounding using the "modern" TZ definition instead of the "contemporary" one? Maybe we should just not test this case then, since this is down to whatever the timezone database for a system contains…? Or am I misunderstanding.

Comment (rok) [reply python/pyarrow/tests/test_compute.py:1945]: I think it's that but I'm not 100% and I really want to be :))

Comment (rok) [reply python/pyarrow/tests/test_compute.py:1945]: Sorry I mislabeled that (switch actual and expected). It appears my rounding in local time is not done correctly.

Comment (rok) [reply python/pyarrow/tests/test_compute.py]: I'm back to thinking Pandas only uses "modern" TZs while date.h uses a historical database.
I've added a [test for this in c++](https://github.com/apache/arrow/blob/e2b6d9f24d37e50bc62e23203e4f6ba6e9275d1f/cpp/src/arrow/compute/kernels/scalar_temporal_test.cc#L1912-L1945) and removed it from Python.

File: cpp/src/arrow/compute/kernels/scalar_temporal_test.cc [hunk 16]
```diff
@@ -1324,6 +1357,592 @@ TEST_F(ScalarTemporalTest, TestTemporalDifferenceZoned) {
...
+
+  CheckScalarUnary("ceil_temporal", unit, times, unit, ceil_1_hours, &round_to_1_hours);
+  CheckScalarUnary("ceil_temporal", unit, times, unit, ceil_2_hours, &round_to_2_hours);
+  CheckScalarUnary("floor_temporal", unit, times, unit, floor_1_hours, &round_to_1_hours);
+  CheckScalarUnary("floor_temporal", unit, times, unit, floor_2_hours, &round_to_2_hours);
+  CheckScalarUnary("round_temporal", unit, times, unit, round_1_hours, &round_to_1_hours);
+  CheckScalarUnary("round_temporal", unit, times, unit, round_2_hours, &round_to_2_hours);
+}
+
+TEST_F(ScalarTemporalTest, TestCeilFloorRoundTemporalKolkata) {
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:1912]: nit: for context, can we note that this test is here to test the a case where `date` (and hence our kernels) uses historical TZ info while Pandas (and possibly other libraries) do not?

Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:1912]: Otherwise I fear confusion when someone in the future looks and wonders why Kolkata was an especially important time zone to test :)

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
Oh, and we need to add this to the python api.rst as well.

File: cpp/src/arrow/compute/kernels/scalar_temporal_test.cc (hunk 16 above)
Comment (rok) [reply cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:1912]: Yeah, that makes sense.

Comment (rok) [reply cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:1912]: [Comment added](https://github.com/apache/arrow/blob/e8f9a621e37fe989e7ae5297886ab7decc069203/cpp/src/arrow/compute/kernels/scalar_temporal_test.cc#L1912-L1948).

It seems reality is [more nuanced](https://www.timeanddate.com/time/zone/india/kolkata) than [Wikipedia states](https://en.wikipedia.org/wiki/Calcutta_Time).

**Review by lidavidm — APPROVED** [review]

**Comment by rok** [comment]
Thanks @lidavidm & @jorisvandenbossche ! :)
I'll make a follow-up for:
* shorthand
* additional timezone handling options (nonexistent, ambiguous)
